rules_bazel/java/java_integration_test.bzl

```diff
@@ -0,0 +1,82 @@
+def _compare_with_goldens_test_impl(ctx):
+    diff_output = ctx.outputs.diff_output
+    check_diff_script = ctx.outputs.check_diff_script
+    input = ctx.attr.api_target
+    srcs = ctx.files.srcs
+    api_name = ctx.attr.name
+    
+    script = """
+    mkdir codegen_tmp
+    unzip -j {input} -d codegen_tmp
+    unzip -j {input_resource_name} -d  codegen_tmp
+    unzip -j {input_test} -d codegen_tmp
+    cd codegen_tmp
+    # Remove unneeded non-Java files, like MANIFEST
+    rm -rf $(find . -type f ! -name "*.java")
+    cd ..
+    diff codegen_tmp test/integration/goldens/{api_name}/ > {diff_output}
+    """.format(
+        diff_output = diff_output.path,
+        input = input[0][JavaInfo].source_jars[0].path,
+        input_resource_name = input[1][JavaInfo].source_jars[0].path,
+        input_test = input[2][JavaInfo].source_jars[0].path,
+        api_name = api_name
+    )
+    ctx.actions.run_shell(
+        inputs = srcs + [
+            input[0][JavaInfo].source_jars[0], 
+            input[1][JavaInfo].source_jars[0], 
+            input[2][JavaInfo].source_jars[0],
+        ],
+        outputs = [diff_output],
+        command = script,
+    )
+
+
+    check_diff_script_content = """
+    cd ${{BUILD_WORKSPACE_DIRECTORY}}
```

Comment:
Can I get some input here? @miraleung @vam-google 

The path of `diff_output` is `bazel-out/k8-fastbuild/bin/test/integration/redis_diff.txt`, but seems we won't get access to `bazel-out` folders in this environment. `${{BUILD_WORKSPACE_DIRECTORY}}` does not work well here, it complains about `HOME not set`. 

Some available paths according to this [doc](https://docs.bazel.build/versions/master/test-encyclopedia.html#initial-conditions)
```
>> PWD: /usr/local/google/home/xiaozhenliu/.cache/bazel/_bazel_xiaozhenliu/96b19089c047889e0621b0e06e850da1/sandbox/linux-sandbox/1/execroot/gapic_generator_java/bazel-out/k8-fastbuild/bin/test/integration/codegen_comparison_script.sh.runfiles/gapic_generator_java

>> TEST_TMPDIR: /usr/local/google/home/xiaozhenliu/.cache/bazel/_bazel_xiaozhenliu/96b19089c047889e0621b0e06e850da1/sandbox/linux-sandbox/1/execroot/gapic_generator_java/_tmp/9ce7325f79eb1f0710ae560ae2ac77ca

>> TEST_WORKSPACE: gapic_generator_java

>> TEST_SRCDIR: /usr/local/google/home/xiaozhenliu/.cache/bazel/_bazel_xiaozhenliu/96b19089c047889e0621b0e06e850da1/sandbox/linux-sandbox/1/execroot/gapic_generator_java/bazel-out/k8-fastbuild/bin/test/integration/
codegen_comparison_script.sh.runfiles
```

Reply:
When the `BUILD_WORKSPACE_DIRECTORY` env variable is not present, in which context do you run your command: `bazel build`, `bazel run`, `bazel test`? I believe `BUILD_WORKSPACE_DIRECTORY` is there only for bazel run.

Reply:
1) In the goldens files rules, we still use `actions.write` [here](https://github.com/googleapis/gapic-generator-java/blob/master/rules_bazel/java/java_diff_test.bzl#L36) I think last time it needs extra `bazel build` because the test runner (code generation) is separate from the goldens comparison step. Please correct me if wrong.  And This case is similar to goldens files and `actions.run_shell` will not work because as an executable rule, we need to return a `[DefaultInfo(executable =..)]`, `actions.write()` is for writing such a shell script which can be executed when using running `bazel test //test/integration:redis`.

2) This `actions.write()` is trying to write a script and return executable. The script will check if the `diff_output.txt` is empty or not: empty means no diff, then test passes, while if the diff file is not empty, fail the test. I'm trying to define a way to tell the test succeed or not.

Now the main question is that when entering the second `actions.write()`, the bazel enters in test mode, it's running in a isolated container, so the generated srcjar or diff file can not be passed in. If we run this rule by `bazel run //test/integration:redis` it will work fine, because in `bazel run` mode, we have access to the `BUILD_WORKSPACE_DIRECTORY ` and all generated files which is located at `bazel-out/k8-fastbuild/bin/test/integration/redis_diff.txt` @miraleung @vam-google

Reply:
This is resolved, the data we want to use in the test rules can be passed by `runfiles` attribute.